### ISSUE TITLE
Add Support for KeyedServices

### ISSUE THREAD

**Issue by HakamFostok** [dotnet/maui#18992]
### Description

`ServiceProvider` in Maui does not support KeyedServices

### Steps to Reproduce

`Microsoft.Maui.Controls.Application.Current?.Handler.MauiContext?.Services.GetRequiredKeyedService<T>(key);`

### Version with bug

8.0.3

### Is this a regression from previous behavior?

No, this is something new

### Last version that worked well

Unknown/Other

### Affected platforms

I was *not* able test on other platforms

### Did you find any workaround?

No

### Relevant log output

```shell
System.InvalidOperationException: 'This service provider doesn't support keyed services.'
```

**Comment by jfversluis** [comment]
Changing this to be a feature request rather than a bug, thanks!

**Comment by Eilon** [comment]
Related PR: https://github.com/dotnet/runtime/pull/89509

**Comment by HakamFostok** [comment]
Just a question, not a promise, do you think that this feature is not complex enough to the level that I can help with implementing it? I would like to help but I am not sure If I really can.
Please let me know if I can help even partially.
Thank you

**Comment by Eilon** [comment]
@HakamFostok good question, I'm not sure. It _could_ be as simple as roughly copying the changes and tests from https://github.com/dotnet/runtime/pull/89509 and applying to MAUI. Or does MAUI even get some of those changes for free due to possibly using the same types?

I'm paging @halter73 and @BrennanConroy who might have some insights about it. But at a glance it doesn't _seem_ too complicated, so @HakamFostok if you want to spend a bit of time on it to see if maybe it's very easy, that could be worthwhile.

**Comment by HakamFostok** [comment]
I know this is very far from complete, but I am not sure if I am on the right path or not, so I really need some guidance,

Please take a look at this commit, whenever you have the time https://github.com/dotnet/maui/commit/7d3ef3f6f43e6cceac9c1f30069a752a6640cce6

I am just trying to get feedback as soon as possible, and just start to put my foot on the path, this is not a full implementation by any means.

Thank you @Eilon @jfversluis

**Comment by Eilon** [comment]
@HakamFostok thanks I also just saw that @mattleibow submitted https://github.com/dotnet/maui/pull/20014.

**Comment by mattleibow** [comment]
I am thinking we don't need to do much as we do use the same service provider from dotnet/runtime. Our issue is that we have a wrapper service because we need another feature. Our wrapper is missing the new interface.

**Comment by HakamFostok** [comment]
Thank you a lot @Eilon.
For me, the most important thing is to see this supported, no matter who does it.
I really believe it is much better to be done by Microsoft itself, so I am very happy to see this happening ❤️ 
I am also glad what I did was almost the same the the PR you mentioned (I did not write UnitTest at the time because I was not sure if I was on the correct path), but now I am very happy to see that the things I did was not completely crazy, that is really make me more confident in my self.

Thank you again all for your great efforts to make .NET and MAUI great products ❤️
I will delete my fork and hopefully, we will see this in .NET 9
